### ISSUE TITLE
Imposm tiles expiry doesn't refresh materialized views and leads to old geometry display at refresh

### ISSUE THREAD

**Issue by flacombe** [openinframap/styles#96]
Following dockerisation tests in #95 it appears that purging tegola tiles isn't enough to refresh tiles when they rely on materialized views.
Imposm3 publishes a log which makes corresponding tiles expired due to a given update but some materialized views could be refreshed hours after.
Then any user asking for a tile that has expired short before in this interval of time will get old geometries that could stay indefinitely (waiting for the next expiration in the area).

Imposm3 could make tiles expired every minutes and refreshing materialized views so often sounds overkill.
Should expired.py script be run less often (let's say every 15 minutes) to let time to views to be refreshed?

We could also wait for https://github.com/go-spatial/tegola/issues/742 to be fixed to continue purging layers that rely on tables every minute and only completely purge materialized layers daily on views refresh.

Looking for comments about this issue, have a nive weekend :)

**Comment by russss** [comment]
Yeah, this is an issue but I haven't managed to work out a good way of fixing it.

Incidentally I'm currently running the materialized view update job every 10 minutes (although it has got slower recently). I also have a job which purges cached tiles older than 14 days for openinframap so there is an upper limit for how long these will persist.

**Comment by flacombe** [comment]
That's good points, thanks you to bring this here.

The only solution I tried is to process expired tiles after materialized views refresh.
In #93, Materialized views are refreshed every 10 minutes (*/10) by impsom3 docker
In #95, expired tiles are processed every 10 minutes with 2 minutes offset (2,12,22,32,42,52) by Tegola docker.

Additionally, I've modified the expired tiles processing a bit to set back the responsibility of clean up in imposm3 docker.
It allows to cleanup tiles log even if Tegola isn't running and let several Tegola dockers consume logs during a 10 minutes time interval without causing any collision between each others.

It would be possible to make the offset between views refresh and expired processing a little longer depending on views refresh performance.

Let me know how do you feel with this

**Comment by russss** [comment]
Yes, this would mostly fix it, although there's still a race condition where an update could come in during that process and be missed. The materialized views refresh currently takes around 1 minute on my system.

However, I currently run the expiry job every minute, and I have plans to speed this up further (and also altering it to purge the downstream Cloudflare cache). I'm not especially keen on slowing this down, especially since the majority of updates don't affect relations.

I have avoided merging your Docker changes so far, because I don't want to do this until I migrate my own setup to Docker (which I will be doing at some point - it's already partly using docker). It's quite possible I'll have to make some other changes when I do that, but I'll try and keep them so you can at least inherit from those containers.

**Comment by flacombe** [comment]
That sounds understandable and I agree it's not the best option to slow down the tile expiry processing with a direct impact on user experience.

It's ok to wait until you completely moved your infrastructure to Docker and inherit. It's already the case as I use several tegola for a single imposm docker.
I hope you'll be able to keep imposm and tegola on two separate images to allow reuse them elsewhere. Let me know.

Currently, I solved the update-while-refreshing materialized views situation by overlapping expiry processing periods : 13 minutes of expiry every 10 minutes, as to catch up on possible update occurred right before or during the mat views refresh.
Maybe inotify + splitting the map between layers with or without mat views could make the expiry processing pretty immediate (among other possibilities)

**Comment by russss** [comment]
Yeah, I was planning to move the expiry process to using inotify eventually - I think that combined with purging the Cloudflare cache (and changing the cache headers so the tiles are cached for a shorter time in browsers) will make things even more responsive.

One possible solution for the relations is to add a trigger to the relation tables which inserts a row into another table whenever a relation is changed, then a script could listen for that, run the materialised view update, then purge the appropriate tiles. This would definitely work, but it adds quite a bit more complexity.

**Comment by flacombe** [comment]
Hello @russss 

It's fine with python script indeed.
Isn't it possible to use the existing https://github.com/openinframap/openinframap/blob/main/schema/refresh_matviews.sql file inside?
It will ease maintenance to split views to be updated from the python script.

**Comment by russss** [comment]
I don't think it's really worth it while it's only two queries, and I don't think there are likely to be any more added. I might delete the SQL file just so everything's in one place.

**Comment by flacombe** [comment]
Schema is completely described in /schema and there can be more mat views ([I have 5](https://github.com/flacombe/openinframap/blob/latest/schema/refresh_matviews.sql) and some other business logic). This could confuse reusers
